icicle/fields/CMakeLists.txt

```diff
@@ -0,0 +1,72 @@
+cmake_minimum_required(VERSION 3.18)
+
+include(../Common.cmake)
+
+set_env()
+set_gpu_env()
+
+project(icicle LANGUAGES CUDA CXX)
+
+# when adding a new curve/field, append its name to the end of this list
+set(SUPPORTED_FIELDS bn254;bls12_381;bls12_377;bw6_761;grumpkin)
+
+set(IS_FIELD_SUPPORTED FALSE)
+set(I 0)
+foreach (SUPPORTED_FIELD ${SUPPORTED_FIELDS})
+  math(EXPR I "${I} + 1")
+  if (FIELD STREQUAL SUPPORTED_FIELD)
+    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -DFIELD_ID=${I}")
+    set(IS_FIELD_SUPPORTED TRUE)
+  endif ()
+endforeach()
+
+if (NOT IS_FIELD_SUPPORTED)
+  message( FATAL_ERROR "The value of FIELD variable: ${FIELD} is not one of the supported fields: ${SUPPORTED_FIELDS}" )
+endif ()
+
+if (G2_DEFINED STREQUAL "ON")
+  set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -DG2_DEFINED=ON")
+endif ()
+
+option(BUILD_TESTS "Build tests" OFF)
+
+if (NOT BUILD_TESTS)
+
+  message(STATUS "Building without tests.")
+
+  add_library(icicle ${FIELD}/extern.cu)
```

Comment:
something not clear to me: here you define icicle as one library but in another library you define it as well as another one.
I think you need to name the libraries better  (maybe here icicle_fields or fields or so) and have icicle link to all of them

Reply:
also note that if multiple libs would depend on fields (probably will happen) then we will end up with multiple symbols so the library would end up larger with duplicate code. I think this should not cause link issues but not ideal. 
For example both curves and polynomials would depend on fields so both would link to fields and we end up with fields twice in the final icicle library. Unless I misunderstood your intention.

One option is to use shared libs. Or think about the structure again.